base/sort.jl

```diff
@@ -338,38 +343,42 @@ function sort!(v::AbstractVector, lo::Int, hi::Int, a::MergeSortAlg, o::Ordering
     return v
 end
 
-function sort!(v::AbstractVector, lo::Int, hi::Int, a::PartialQuickSort{Int},
-               o::Ordering)
-    @inbounds while lo < hi
-        hi-lo <= SMALL_THRESHOLD && return sort!(v, lo, hi, SMALL_ALGORITHM, o)
-        j = partition!(v, lo, hi, o)
-        if j >= a.k
-            # we don't need to sort anything bigger than j
-            hi = j-1
-        elseif j-lo < hi-j
-            # recurse on the smaller chunk
-            # this is necessary to preserve O(log(n))
-            # stack space in the worst case (rather than O(n))
-            lo < (j-1) && sort!(v, lo, j-1, a, o)
-            lo = j+1
-        else
-            (j+1) < hi && sort!(v, j+1, hi, a, o)
-            hi = j-1
-        end
-    end
-    return v
-end
-
-
-function sort!{T<:OrdinalRange}(v::AbstractVector, lo::Int, hi::Int, a::PartialQuickSort{T},
+## TODO: When PartialQuickSort is parameterized by an Int, this version of sort
+##       has one less comparison per loop than the version below, but enabling
+##       it causes return type inference to fail for sort/sort! (#12833)
+##
+# function sort!(v::AbstractVector, lo::Int, hi::Int, a::PartialQuickSort{Int},
```

Comment:
This didn't fix the overallocations: https://github.com/JuliaLang/julia/issues/12833#issuecomment-135845045